sdks/java/io/hadoop-file-system/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java

```diff
@@ -313,7 +328,7 @@ protected HadoopResourceId matchNewResource(String singleResourceSpec, boolean i
 
   @Override
   protected String getScheme() {
-    return fileSystem.getScheme();
+    return "hdfs";
```

Comment:
I'm little bit concerned about this. We should probably have more registrars for different schemas. At least for `webhdfs`, which is commonly used.

Reply:
The original intent was for a different scheme to be returned for each different hdfs configuration so a user if they wanted could register hdfs1, hdfs2, webhdfs, ...

All the registration logic is happening based in https://github.com/apache/beam/blob/553c56075b0fdcbc1a3d37d8b708697f1ee1f5de/sdks/java/io/hadoop-file-system/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemRegistrar.java#L42 and each HadoopFileSystem instance is mean to be related to a different configuration.

This way when they are writing their pipeline they would write `Read.from("hdfs1://...")`

Reply:
Got the idea. Anyway don't think, that `hdfs\d+` is a valid scheme, that hadoop understands.

If we want to support multiple clusters, using the native protocol, what matters is an authority section of the URL. I think we can introduce some kind of `AuthorityAwareFileSystem`, that would be just a container for the HadoopFileSystems with the same scheme and decide based on the authority part, which filesystem should be used.

There are some things I'm not sure about.
1) What should we do in case there are multiple configs for same scheme + authority? I think failing fast should be the best option here.
2) We are duplicating functionality already provided by Hadoop. This is an equivalent of specifying multiple `nameservice` entries in the same config. Do we want to do this, because it is kind of an unexpected API for an user coming from the Hadoop world?

Reply:
I don't think we need to provide an `AuthorityAwareFileSystem` in Beam. As long as the HadoopFileSystem knows how to handle multiple authorities by selecting the appropriate configuration for it, we are good.

For the other schemes that HDFS currently supports such as webhdfs, ftp, ..., they should all be registered as separate HadoopFileSystem instances.

If Hadoop starts to support arbitrary schemes, the hdfs1, hdfs2, ... is meant to represent the case where you want to access the same authority but with different settings which is why I agree with failing fast when multiple configurations provide the same scheme+authority.

Reply:
If we want to do it this way, it would be a good idea to allow user to set only a single configuration, which would be a breaking change to the API, but we don't allow user to use it anyway right now.

```
    checkArgument(
        configurations.size() <= 1,
        String.format(
            "The %s currently only supports at most a single Hadoop configuration.",
            HadoopFileSystemRegistrar.class.getSimpleName()));
```

This way we wouldn't have to deal with multiple `fs.defaultFS` definitions and other possible config option conflicts.

It is probably less flexible, but we should have either this, or force user to specify a config per cluster.

WDYT?

Reply:
I think it is ok if there are multiple configurations and we choose which one based upon authority within a single HadoopFileSystem instance. Any settings that a user puts down that are conflicting are up to them to resolve. We or Hadoop can provide helpful errors if they do this but it is really up to them to have valid hadoop configurations.